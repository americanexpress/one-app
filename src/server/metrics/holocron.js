/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { createCounter } from './counters';
import { createGauge } from './gauges';

import createMetricNamespace from './create-metric-namespace';

const holocronNamespace = createMetricNamespace('holocron');

createCounter({
  name: holocronNamespace('module_map_poll', 'total'),
  help: 'total times the module map has been polled',
});

createCounter({
  name: holocronNamespace('module_map_updated', 'total'),
  help: 'total times the module map has been updated',
});

createCounter({
  name: holocronNamespace('module_map_poll_restarted', 'total'),
  help: 'total times the module map poll has been restarted',
});

createGauge({
  name: holocronNamespace('module_map_poll_wait', 'seconds'),
  help: 'delay until the next poll of the module map',
});

createGauge({
  name: holocronNamespace('module_map_poll_consecutive_errors', 'total'),
  help: 'how many times has the server encountered errors polling the module map',
});

export default holocronNamespace.getMetricNames();
