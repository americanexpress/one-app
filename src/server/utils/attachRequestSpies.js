/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import https from 'node:https';
import http from 'node:http';
import url from 'node:url';
import onFinished from 'on-finished';

export function attachSpy(obj, methodName, spy) {
  const origMethod = obj[methodName];

  if (typeof origMethod !== 'function') {
    throw new TypeError(`${methodName} is not a function`);
  }

  if (typeof spy !== 'function') {
    throw new TypeError(`spy must be a function (was "${typeof spy}")`);
  }

  // we're monkeypatching, we need to reassign a property of the obj argument
  // eslint-disable-next-line no-param-reassign -- see above comment
  obj[methodName] = function spiedMethod(...args) {
    let returnValue;
    let originalCalled = false;
    const callOriginal = () => {
      originalCalled = true;
      returnValue = origMethod.apply(obj, args);
      return returnValue;
    };

    spy([...args], callOriginal);

    if (!originalCalled) {
      callOriginal();
    }

    return returnValue;
  };
}

function buildUrlObject(options, defaultProtocol) {
  // TODO: url.parse is deprecated, use new URL() instead
  // this is not a 1:1 replacement and will require changes.
  // Currently the parsed url is used to provide href to
  // loggers outgoingRequestEndSpy function.
  // https://github.com/americanexpress/one-app/blob/main/src/server/utils/logging/setup.js#L42
  // https://nodejs.org/api/url.html#urlparseurlstring-parsequerystring-slashesdenotehost
  const parsedPath = url.parse(options.path || '/');
  const protocol = options.protocol || `${defaultProtocol}:`;
  const urlObject = {
    auth: options.auth,
    hostname: options.hostname || options.host || 'localhost',
    protocol,
    port: options.port || (protocol === 'http:' ? 80 : 443),
    hash: parsedPath.hash,
    path: parsedPath.path,
    pathname: parsedPath.pathname,
    query: parsedPath.query,
    search: parsedPath.search,
  };
  if (
    (protocol === 'http:' && urlObject.port === 80)
    || (protocol === 'https:' && urlObject.port === 443)
  ) {
    urlObject.port = undefined;
  }
  return urlObject;
}

function parseUrl(options, defaultProtocol) {
  // some data is not stored in the clientRequest, have to duplicate some parsing
  // adapted from https://github.com/nodejs/node/blob/894203dae39c7f1f36fc6ba126bb5d782d79b744/lib/_http_client.js#L22
  if (typeof options === 'string') {
    return url.parse(options);
  }

  return url.parse(url.format(buildUrlObject(options, defaultProtocol)));
}

function httpSpy(defaultProtocol, requestSpy, socketCloseSpy) {
  return (args, callOriginal) => {
    const options = args[0];
    const clientRequest = callOriginal();
    const parsedUrl = parseUrl(options, defaultProtocol);

    requestSpy(clientRequest, parsedUrl);

    if (socketCloseSpy) {
      onFinished(clientRequest, () => socketCloseSpy(clientRequest, parsedUrl));
    }
  };
}

export default function attachRequestSpies(requestSpy, socketCloseSpy) {
  if (typeof requestSpy !== 'function') {
    throw new TypeError(`requestSpy must be a function (was "${typeof requestSpy}")`);
  }

  if (socketCloseSpy && (typeof socketCloseSpy !== 'function')) {
    throw new TypeError(
      `socketCloseSpy must be function if provided (was "${typeof socketCloseSpy}")`);
  }

  attachSpy(https, 'request', httpSpy('https', requestSpy, socketCloseSpy));
  attachSpy(http, 'request', httpSpy('http', requestSpy, socketCloseSpy));
}
