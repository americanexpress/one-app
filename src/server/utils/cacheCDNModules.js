/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import path from 'path';
import fs from 'fs';
import chalk from 'chalk';

export const getUserHomeDirectory = () => process.env.HOME || process.env.USERPROFILE;
export const cacheFileName = '.one-app-module-cache';
export const moduleCacheFileName = '.one-app-module-map-cache';
export const oneAppDirectoryName = '.one-app';
export const oneAppDirectoryPath = path.join(getUserHomeDirectory(), oneAppDirectoryName);
export const oneAppModuleCachePath = path.join(oneAppDirectoryPath, cacheFileName);

// show cache size and how to delete info on start
export const showCacheInfo = () => {
  fs.stat(oneAppModuleCachePath, (error, stats) => {
    if (error) {
      console.error('There was error checking file stat', error);
    } else {
      const fileSizeOnMB = stats.size / (1024 * 1024); // bytes to mb
      const message = `File size of ${cacheFileName}: ${chalk.bold.greenBright(fileSizeOnMB.toFixed(2), 'MB')}`;
      const separator = '*'.repeat(message.length);
      console.log(chalk.bold.cyanBright(separator));
      console.log(chalk.bold.redBright('CACHE INFORMATION'));
      console.log(message);
      console.log(`To delete cache, please run \n  ${chalk.bold.redBright('  rm ', oneAppModuleCachePath)}`);
      console.log(chalk.bold.cyanBright(separator));
    }
  });
};

// setup folder and file
export const setupCacheFile = () => {
  fs.mkdir(oneAppDirectoryPath, { recursive: true }, (error) => {
    if (error) {
      console.error(`There was error creating ${oneAppDirectoryName} directory`);
    } else {
      console.log(`Successfully created ${oneAppDirectoryPath}`);
      console.log(`Creating ${cacheFileName}`);
      try {
        fs.writeFileSync(oneAppModuleCachePath, JSON.stringify('{}'));
        console.log(`${cacheFileName} created successfully on ${oneAppModuleCachePath}`);
      } catch (err) {
        console.error(`Error creating ${cacheFileName} on ${oneAppModuleCachePath}, \n${err}`);
      }
    }
  });
};

// gets cached module from ~/.one-app/.one-app-module-cache
export const getCachedModules = () => {
  let hasCachedFile = false;
  if (fs.existsSync(oneAppModuleCachePath)) {
    hasCachedFile = true;
  } else {
    setupCacheFile();
  }
  if (hasCachedFile) {
    try {
      showCacheInfo();
      const cachedContent = fs.readFileSync(oneAppModuleCachePath, 'utf8');
      return JSON.parse(cachedContent);
    } catch (err) {
      console.error('Could not parse JSON content', err);
    }
  }
  return {};
};

let timerId = null;

export const writeToCache = (content, delay = 500) => {
  // added debounce
  clearTimeout(timerId);
  timerId = setTimeout(() => {
    fs.writeFile(oneAppModuleCachePath, JSON.stringify(content, null, 2), (error) => {
      if (error) {
        console.log(`There was an error updating content \n ${error}`);
      }
    });
    timerId = null;
  }, delay);
};

export const removeDuplicatedModules = (url, cachedModules, moduleNames) => {
  const matchingModule = moduleNames.find((moduleName) => url.match(new RegExp(`\\b\\/${moduleName}\\/\\b`)));

  const updatedCachedModules = cachedModules;
  Object.keys(updatedCachedModules).forEach((cachedModuleKey) => {
    if (cachedModuleKey.match(new RegExp(`\\b\\/${matchingModule}\\/\\b`))) {
      delete updatedCachedModules[cachedModuleKey];
      console.log(`Deleted ${cachedModuleKey} from cache`);
    }
  });
  return updatedCachedModules;
};
