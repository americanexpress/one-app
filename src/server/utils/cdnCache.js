/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import path from 'node:path';
import fs, { promises as fsPromises } from 'node:fs';
import chalk from 'chalk';

export const getUserHomeDirectory = () => process.env.HOME || process.env.USERPROFILE;
export const cacheFileName = '.one-app-module-cache';
export const oneAppDirectoryName = '.one-app';
export const oneAppDirectoryPath = path.join(getUserHomeDirectory(), oneAppDirectoryName);
export const oneAppModuleCachePath = path.join(oneAppDirectoryPath, cacheFileName);

// show cache size and how to delete info on start
export const showCacheInfo = async () => {
  try {
    const stats = await fsPromises.stat(oneAppModuleCachePath);
    const fileSizeInMB = stats.size / (1024 * 1024); // bytes to mb
    const separator = '*'.repeat(50);
    console.info(chalk.bold.cyanBright(separator));
    console.info(chalk.bold.cyanBright('CACHE INFORMATION'));
    console.info('File size of %s: %s', cacheFileName, chalk.bold.greenBright(fileSizeInMB.toFixed(2), 'MB'));
    console.info('To clear the cache, please delete this file:');
    console.info('    %s', chalk.bold.redBright(path.join('~', oneAppDirectoryName, cacheFileName)));
    console.info(chalk.bold.cyanBright(separator));
  } catch (error) {
    console.error('There was error checking file stat', error);
  }
};

// setup folder and file
export const setupCacheFile = async () => {
  try {
    await fsPromises.mkdir(oneAppDirectoryPath, { recursive: true });
    console.log(`Successfully created ${oneAppDirectoryPath}`);
    console.log(`Creating ${cacheFileName}`);
    try {
      await fsPromises.writeFile(oneAppModuleCachePath, JSON.stringify({}));
      console.log(`${cacheFileName} created successfully on ${oneAppModuleCachePath}`);
    } catch (error) {
      console.error(`Error creating ${cacheFileName} on ${oneAppModuleCachePath}, \n${error}`);
    }
  } catch (error) {
    console.error(`There was error creating ${oneAppDirectoryName} directory`);
  }
};

// gets cached module from ~/.one-app/.one-app-module-cache
export const getCachedModuleFiles = () => {
  if (!fs.existsSync(oneAppModuleCachePath)) {
    setupCacheFile();
    return {};
  }

  try {
    showCacheInfo();
    const cachedContent = fs.readFileSync(oneAppModuleCachePath, 'utf8');
    return JSON.parse(cachedContent);
  } catch (error) {
    console.error('Could not parse JSON content', error);
    return {};
  }
};

let timerId = null;

export const writeToCache = (content, delay = 500) => {
  // added debounce
  clearTimeout(timerId);
  timerId = setTimeout(() => {
    fs.writeFile(oneAppModuleCachePath, JSON.stringify(content, null, 2), (error) => {
      if (error) {
        console.log(`There was an error updating content \n ${error}`);
      }
    });
    timerId = null;
  }, delay);
};

const stripVersion = (url) => {
  const parts = url.split('/');
  parts.splice(-2, 1);
  return parts.join('/');
};

export const removeExistingEntryIfConflicting = (url, cachedModuleFiles) => {
  const updatedCachedModules = { ...cachedModuleFiles };
  const strippedUrl = stripVersion(url);

  const matchingModule = Object.keys(cachedModuleFiles)
    .find((cachedUrl) => stripVersion(cachedUrl) === strippedUrl);

  if (matchingModule && matchingModule !== url) {
    delete updatedCachedModules[matchingModule];
  }
  return updatedCachedModules;
};
