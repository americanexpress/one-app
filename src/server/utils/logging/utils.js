/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { STATUS_CODES } from 'http';

import chalk from 'chalk';

const coloredLevels = {
  error: chalk.red('error'),
  warn: chalk.yellow('warn'),
  log: chalk.blue('log'),
  info: chalk.gray('info'),
};

function printStatusCode(obj) {
  const { request: { statusCode } } = obj;

  if (statusCode === null) {
    return chalk.black.bgRed(statusCode);
  }

  if (statusCode < 300) {
    // 100s, 200s
    return chalk.green(statusCode);
  }
  if (statusCode < 400) {
    // 300s
    return chalk.blue(statusCode);
  }
  if (statusCode < 500) {
    // 400s
    return chalk.yellow(statusCode);
  }
  // 500s
  return chalk.red(statusCode);
}

function printStatusMessage(obj) {
  const { request: { statusText, statusCode } } = obj;

  return statusText || STATUS_CODES[statusCode] || chalk.black.bgRed('Timed Out?');
}

function printDurationTime(obj) {
  const { request: { timings: { duration } } } = obj;
  if (duration < 100) {
    return chalk.green(duration);
  }
  if (duration < 1e3) {
    return chalk.yellow(duration);
  }
  if (duration < 6e3) {
    return chalk.red(duration);
  }
  return chalk.black.bgRed(duration);
}

export {
  coloredLevels,
  printStatusCode,
  printStatusMessage,
  printDurationTime,
};
