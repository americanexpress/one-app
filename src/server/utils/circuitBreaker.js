/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

/* global BigInt */

import { promisify } from 'util';
import CircuitBreaker from 'opossum';
import PrometheusMetrics from 'opossum-prometheus';
import { register } from 'prom-client';
import { composeModules, getModule } from 'holocron';
import { getServerStateConfig } from './stateConfig';

const immediate = promisify(setImmediate);
const { rootModuleName } = getServerStateConfig();
let eventLoopLagThreshold = 30;

export const setEventLoopLagThreshold = (n) => {
  eventLoopLagThreshold = Number(n) || 30;
};

export const getEventLoopLagThreshold = () => eventLoopLagThreshold;

const getModuleData = async ({ dispatch, modules }) => {
  await dispatch(composeModules(modules));
  return false;
};

const options = {
  timeout: false, // Do not use a timeout
  errorThresholdPercentage: 1, // When 1% of requests fail, open the circuit
  resetTimeout: 10e3, // After 10 seconds, try again.
};

const breaker = new CircuitBreaker(getModuleData, options);
// Just need to connect opossum to prometheus
// eslint-disable-next-line no-unused-vars
const metrics = new PrometheusMetrics(breaker, register);

breaker.fallback(() => true);

breaker.healthCheck(async () => {
  if (!getModule(rootModuleName)) return;
  const start = process.hrtime.bigint();
  await immediate();
  const end = process.hrtime.bigint();
  const diff = (end - start) / BigInt(1e6);
  // Open the circuit if event loop lag is greater than threshold
  if (diff > eventLoopLagThreshold) {
    throw new Error(`Opening circuit, event loop lag (${diff}ms) is > eventLoopLagThreshold (${eventLoopLagThreshold}ms)`);
  }
}, 100);

export default breaker;
