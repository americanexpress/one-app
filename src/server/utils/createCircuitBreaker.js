/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { monitorEventLoopDelay } from 'perf_hooks';
import CircuitBreaker from 'opossum';
import { getModule } from 'holocron';
import { getServerStateConfig } from './stateConfig';
import { registerCircuitBreaker } from '../metrics/circuit-breaker';

const { rootModuleName } = getServerStateConfig();
let eventLoopDelayThreshold = 250;

export const setEventLoopDelayThreshold = (n) => {
  eventLoopDelayThreshold = Number(n) || 250;
};

export const getEventLoopDelayThreshold = () => eventLoopDelayThreshold;

const options = {
  // Do not use a timeout
  timeout: false,
  // When 10% of requests fail, open the circuit
  errorThresholdPercentage: 10,
  // After 10 seconds, try again.
  resetTimeout: 10e3,
};

const eventLoopDelayHistogram = monitorEventLoopDelay();
eventLoopDelayHistogram.enable();

const checkMaxEventLoopDelay = async () => {
  // Return if root module is not loaded, as that is where threshold is configured
  if (!getModule(rootModuleName)) return;
  // Get event loop delay in milliseconds (from nanoseconds)
  const maxEventLoopDelay = eventLoopDelayHistogram.max / 1e6;
  // Open the circuit if event loop delay is greater than threshold
  if (maxEventLoopDelay > eventLoopDelayThreshold) {
    eventLoopDelayHistogram.reset();
    throw new Error(`Opening circuit, event loop delay (${maxEventLoopDelay}ms) is > eventLoopDelayThreshold (${eventLoopDelayThreshold}ms)`);
  }
};

const createCircuitBreaker = (asyncFuntionThatMightFail) => {
  // asyncFuntionThatMightFail should return false to indicate fallback is not needed
  const breaker = new CircuitBreaker(asyncFuntionThatMightFail, options);
  // Fallback returns true to indicate fallback behavior is needed
  breaker.fallback(() => true);
  // Check the max event loop delay every 500ms
  breaker.healthCheck(checkMaxEventLoopDelay, 500);
  // Track circuit breaker metrics
  registerCircuitBreaker(breaker);
  return breaker;
};

export default createCircuitBreaker;
