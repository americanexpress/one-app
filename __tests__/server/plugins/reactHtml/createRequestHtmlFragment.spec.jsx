/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import url from 'url';
import { browserHistory, matchPromise } from '@americanexpress/one-app-router';
import { Map as iMap, fromJS } from 'immutable';
import { composeModules } from 'holocron';
// eslint-disable-next-line import/named -- getBreaker is only added in the mock
import { getBreaker } from '../../../../src/server/utils/createCircuitBreaker';

import * as reactRendering from '../../../../src/server/utils/reactRendering';

jest.spyOn(console, 'error').mockImplementation(() => 0);

jest.mock('@americanexpress/one-app-router', () => ({
  ...jest.requireActual('@americanexpress/one-app-router'),
  matchPromise: jest.fn(),
}));

jest.mock('holocron', () => ({
  composeModules: jest.fn(() => 'composeModules'),
  getModule: () => () => 0,
}));

jest.mock('../../../../src/server/utils/createCircuitBreaker', () => {
  const breaker = jest.fn();
  const mockCreateCircuitBreaker = (asyncFunctionThatMightFail) => {
    breaker.fire = jest.fn(async (...args) => asyncFunctionThatMightFail(...args));
    return breaker;
  };
  mockCreateCircuitBreaker.getBreaker = () => breaker;
  return mockCreateCircuitBreaker;
});

const renderForStringSpy = jest.spyOn(reactRendering, 'renderForString');
const renderForStaticMarkupSpy = jest.spyOn(reactRendering, 'renderForStaticMarkup');

describe('createRequestHtmlFragment', () => {
  let req;
  let res;
  let createRoutes;
  const dispatch = jest.fn((x) => x);
  const getState = jest.fn(() => fromJS({
    rendering: iMap({}),
  }));

  beforeAll(() => {
    matchPromise.mockImplementation(({ routes, location }) => Promise.resolve({
      redirectLocation: undefined,
      renderProps: {
        routes,
        components: [() => 'hi'],
        location: url.parse(location),
        router: browserHistory,
        params: {
          hi: 'there?',
        },
      },
    }));
  });

  beforeEach(() => {
    jest.clearAllMocks();
    req = jest.fn();
    req.log = { error: jest.fn() };
    req.headers = {};

    res = jest.fn();
    res.status = jest.fn(() => res);
    res.sendStatus = jest.fn(() => res);
    res.redirect = jest.fn(() => res);
    res.end = jest.fn(() => res);
    res.code = jest.fn();
    req.url = 'http://example.com/request';
    req.store = { dispatch, getState };

    createRoutes = jest.fn(() => [{ path: '/', moduleName: 'root' }]);

    renderForStringSpy.mockClear();
    renderForStaticMarkupSpy.mockClear();
    req.log.error.mockClear();
  });

  const requireCreateRequestHtmlFragment = (...args) => require(
    '../../../../src/server/plugins/reactHtml/createRequestHtmlFragment'
  ).default(...args);

  it('should preload data for matched route components', async () => {
    expect.assertions(4);

    await requireCreateRequestHtmlFragment(req, res, { createRoutes });

    expect(getBreaker().fire).toHaveBeenCalled();
    expect(composeModules).toHaveBeenCalled();
    expect(composeModules.mock.calls[0][0]).toMatchSnapshot();
    expect(dispatch).toHaveBeenCalledWith('composeModules');
  });

  it('should add app HTML to the request object', async () => {
    expect.assertions(4);

    await requireCreateRequestHtmlFragment(req, res, { createRoutes });

    expect(createRoutes).toHaveBeenCalledWith(req.store);
    expect(renderForStringSpy).toHaveBeenCalled();
    expect(renderForStaticMarkupSpy).not.toHaveBeenCalled();
    expect(typeof req.appHtml).toBe('string');
  });

  it('should add app HTML as static markup to the request object when scripts are disabled', async () => {
    expect.assertions(4);

    getState.mockImplementationOnce(() => fromJS({
      rendering: iMap({ disableScripts: true }),
    }));

    await requireCreateRequestHtmlFragment(req, res, { createRoutes });

    expect(createRoutes).toHaveBeenCalledWith(req.store);
    expect(renderForStringSpy).not.toHaveBeenCalled();
    expect(renderForStaticMarkupSpy).toHaveBeenCalled();
    expect(typeof req.appHtml).toBe('string');
  });

  it('should set the custom HTTP status', async () => {
    expect.assertions(3);

    createRoutes = jest.fn(() => [{ path: '/', httpStatus: 400 }]);
    await requireCreateRequestHtmlFragment(req, res, { createRoutes });

    expect(createRoutes).toHaveBeenCalledWith(req.store);
    expect(res.code).toHaveBeenCalledWith(400);
    expect(typeof req.appHtml).toBe('string');
  });

  it('does not generate HTML when no route is matched', async () => {
    expect.assertions(4);

    matchPromise.mockImplementationOnce(() => ({
      redirectLocation: undefined,
      // omit renderProps
    }));

    await requireCreateRequestHtmlFragment(req, res, { createRoutes });

    expect(req.log.error).toHaveBeenCalledWith('error creating request HTML fragment for %s', 'http://example.com/request', expect.any(Error));
    expect(res.code).toHaveBeenCalledWith(404);
    expect(createRoutes).toHaveBeenCalledWith(req.store);
    expect(req.appHtml).toBe(undefined);
  });

  it('redirects when a relative redirect route is matched', async () => {
    expect.assertions(3);

    matchPromise.mockImplementationOnce(() => ({
      redirectLocation: {
        pathname: '/redirect',
        search: '',
      },
    }));

    await requireCreateRequestHtmlFragment(req, res, { createRoutes });

    expect(createRoutes).toHaveBeenCalledWith(req.store);
    expect(res.redirect).toHaveBeenCalledWith(302, '/redirect');
    expect(req.appHtml).toBe(undefined);
  });

  it('redirects when an absolute redirect route is matched', async () => {
    expect.assertions(3);

    matchPromise.mockImplementationOnce(() => ({
      redirectLocation: {
        state: url.parse('https://example.com/redirect'),
      },
    }));

    await requireCreateRequestHtmlFragment(req, res, { createRoutes });

    expect(createRoutes).toHaveBeenCalledWith(req.store);
    expect(res.redirect).toHaveBeenCalledWith(302, 'https://example.com/redirect');
    expect(req.appHtml).toBe(undefined);
  });

  it('should catch any errors and call the next middleware', async () => {
    expect.assertions(2);

    const createRoutesError = new Error('failed to create routes');
    const brokenCreateRoutes = () => { throw createRoutesError; };

    await requireCreateRequestHtmlFragment(req, res, { createRoutes: brokenCreateRoutes });

    expect(req.log.error).toHaveBeenCalled();
    expect(req.log.error.mock.calls[0]).toEqual(['error creating request HTML fragment for %s', 'http://example.com/request', createRoutesError]);
  });

  it('should use a circuit breaker', async () => {
    expect.assertions(5);

    await requireCreateRequestHtmlFragment(req, res, { createRoutes });

    expect(getBreaker().fire).toHaveBeenCalled();
    expect(composeModules).toHaveBeenCalled();
    expect(renderForStringSpy).toHaveBeenCalled();
    expect(renderForStaticMarkupSpy).not.toHaveBeenCalled();
    expect(req.appHtml).toBe('hi');
  });

  it('should fall back when the circuit opens', async () => {
    expect.assertions(4);
    const breaker = getBreaker();
    breaker.fire.mockReturnValueOnce({ fallback: true });

    await requireCreateRequestHtmlFragment(req, res, { createRoutes });

    expect(getBreaker().fire).toHaveBeenCalled();
    expect(renderForStringSpy).not.toHaveBeenCalled();
    expect(renderForStaticMarkupSpy).not.toHaveBeenCalled();
    expect(req.appHtml).toBe('');
  });

  it('should redirect instead of rendering when the circuit breaker returns a redirect', async () => {
    expect.assertions(4);
    composeModules.mockImplementationOnce(() => {
      const error = new Error('An error that redirects');
      error.abortComposeModules = true;
      error.redirect = { status: 302, url: 'https://example.com' };
      throw error;
    });
    await requireCreateRequestHtmlFragment(req, res, { createRoutes });
    expect(res.redirect).toHaveBeenCalledWith(302, 'https://example.com');
    expect(getBreaker().fire).toHaveBeenCalled();
    expect(renderForStringSpy).not.toHaveBeenCalled();
    expect(renderForStaticMarkupSpy).not.toHaveBeenCalled();
  });

  it('should default to a 302 redirect', async () => {
    expect.assertions(1);
    composeModules.mockImplementationOnce(() => {
      const error = new Error('An error that redirects');
      error.abortComposeModules = true;
      error.redirect = { url: 'https://example.com' };
      throw error;
    });
    await requireCreateRequestHtmlFragment(req, res, { createRoutes });
    expect(res.redirect).toHaveBeenCalledWith(302, 'https://example.com');
  });

  it('should rethrow if the error does not contain a redirect', async () => {
    expect.assertions(4);
    composeModules.mockImplementationOnce(() => { throw new Error('An error that does not redirect'); });
    await requireCreateRequestHtmlFragment(req, res, { createRoutes });
    expect(res.redirect).not.toHaveBeenCalled();
    expect(getBreaker().fire).toHaveBeenCalled();
    expect(renderForStringSpy).not.toHaveBeenCalled();
    expect(renderForStaticMarkupSpy).not.toHaveBeenCalled();
  });

  it('should not use the circuit breaker for partials', async () => {
    expect.assertions(5);

    getState.mockImplementationOnce(() => fromJS({
      rendering: {
        renderPartialOnly: true,
      },
    }));

    await requireCreateRequestHtmlFragment(req, res, { createRoutes });

    expect(getBreaker().fire).not.toHaveBeenCalled();
    expect(composeModules).toHaveBeenCalled();
    expect(renderForStringSpy).not.toHaveBeenCalled();
    expect(renderForStaticMarkupSpy).toHaveBeenCalled();
    expect(req.appHtml).toBe('hi');
  });

  it('should not use the circuit breaker when scripts are disabled', async () => {
    expect.assertions(5);

    getState.mockImplementationOnce(() => fromJS({
      rendering: {
        disableScripts: true,
      },
    }));

    await requireCreateRequestHtmlFragment(req, res, { createRoutes });

    expect(getBreaker().fire).not.toHaveBeenCalled();
    expect(composeModules).toHaveBeenCalled();
    expect(renderForStringSpy).not.toHaveBeenCalled();
    expect(renderForStaticMarkupSpy).toHaveBeenCalled();
    expect(req.appHtml).toBe('hi');
  });

  it('should not use the circuit breaker when rendering text only', async () => {
    expect.assertions(5);

    getState.mockImplementationOnce(() => fromJS({
      rendering: {
        renderTextOnly: true,
        renderTextOnlyOptions: { htmlTagReplacement: '', allowedHtmlTags: [] },
      },
    }));

    await requireCreateRequestHtmlFragment(req, res, { createRoutes });

    expect(getBreaker().fire).not.toHaveBeenCalled();
    expect(composeModules).toHaveBeenCalled();
    expect(renderForStringSpy).not.toHaveBeenCalled();
    expect(renderForStaticMarkupSpy).toHaveBeenCalled();
    expect(req.appHtml).toBe('hi');
  });
});
