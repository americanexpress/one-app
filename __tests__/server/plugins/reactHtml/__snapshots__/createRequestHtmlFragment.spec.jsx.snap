// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRequestHtmlFragment should preload data for matched route components 1`] = `
[
  {
    "name": "root",
    "props": {
      "location": Url {
        "auth": null,
        "hash": null,
        "host": "example.com",
        "hostname": "example.com",
        "href": "http://example.com/request",
        "path": "/request",
        "pathname": "/request",
        "port": null,
        "protocol": "http:",
        "query": null,
        "search": null,
        "slashes": true,
      },
      "params": {
        "hi": "there?",
      },
      "route": {
        "moduleName": "root",
        "path": "/",
      },
      "router": {
        "createHref": [Function],
        "createKey": [Function],
        "createLocation": [Function],
        "createPath": [Function],
        "getCurrentLocation": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "listenBefore": [Function],
        "push": [Function],
        "replace": [Function],
        "transitionTo": [Function],
      },
      "routes": [
        {
          "moduleName": "root",
          "path": "/",
        },
      ],
    },
  },
]
`;
